### PR TITLE
#1268 PHP 7.4 support

### LINKED ISSUE #1268: PHP 7.4 support
When do you plan to support PHP 7.4?
After GA or earlier?
I'd like to know when to expect it, so I can start using PHP 7.4 as soon as possible.

Thanks for answer

### PR DESCRIPTION
**Pull request by ddziaduch**
This fixes #1268

### CODE REVIEW

File: src/PhpSpec/Matcher/ScalarMatcher.php [hunk 1]
```diff
@@ -127,6 +127,10 @@ private function getCheckerName(string $name)
             return false;
         }
 
+        if (PHP_VERSION_ID >= 70400 && $name === 'real') {
```
Comment (stof) [inline src/PhpSpec/Matcher/ScalarMatcher.php]: This is a BC break, as specs will suddenly stop passing. I suggest we trigger a deprecation notice instead (we can even do it for all PHP versions) and switch to `is_float` internally (as that's what the alias is internally).

**Comment by stof** [comment]
To be sure that we support PHP 7.4, we would also need to add some Behat scenarios (running only on PHP 7.4+) that would use phpspec with classes using typed properties (we should check it for both the specced class and the collaborator class).

**Comment by drupol** [comment]
Reading this thread makes me think to an extension I've been working on lately.

https://github.com/friends-of-phpspec/phpspec-requires

It let user `requires` a specific PHP version for tests.

Feel free to give some feedback.

**Comment by stof** [comment]
I don't think using third-party phpspec extensions in the suite of phpspec itself is a wise idea (using a tool to test itself is already enough to create weird case sometimes, but relying on third-party extensions becomes really complex to manage, as the extension will never react to PHPSpec changes before they are done...)

**Comment by stof** [comment]
btw, if my suggestion is followed, most of the checks will become unnecessary.

**Comment by drupol** [comment]
> I don't think using third-party phpspec extensions in the suite of phpspec itself is a wise idea (using a tool to test itself is already enough to create weird case sometimes, but relying on third-party extensions becomes really complex to manage, as the extension will never react to PHPSpec changes before they are done...)

I definitely agree and suggesting to use an extension was not my intention.

File: src/PhpSpec/Matcher/ScalarMatcher.php (hunk 1 above)
Comment (ddziaduch) [reply src/PhpSpec/Matcher/ScalarMatcher.php]: Sounds like a great idea !

Comment (ddziaduch) [reply src/PhpSpec/Matcher/ScalarMatcher.php]: I have just added `trigger_error()` with the deprecation but it is catches by PhpSpec so the specs will still fail

File: src/PhpSpec/Matcher/ScalarMatcher.php [hunk 2]
```diff
@@ -127,6 +127,10 @@ private function getCheckerName(string $name)
             return false;
         }
 
+        if ($name == 'beReal') {
+            trigger_error('`beReal` matcher is deprecated, please use `beFloat`', E_USER_DEPRECATED);
```
Comment (stof) [inline src/PhpSpec/Matcher/ScalarMatcher.php]: use `@trigger_error('...', E_USER_DEPRECATED);` to follow the opt-in deprecation reporting that the Symfony community defined (after breaking the whole ecosystem in 2.7.0 when trying to use it without silencing, for similar reasons than what you have here)

Comment (ddziaduch) [reply src/PhpSpec/Matcher/ScalarMatcher.php]: Sorry, but how this is supposed to work? I trigger notice and immediately silence it. How the users will know that this matcher is deprecated?

Comment (ddziaduch) [reply src/PhpSpec/Matcher/ScalarMatcher.php]: @stof OK I found something about it

Comment (stof) [reply src/PhpSpec/Matcher/ScalarMatcher.php]: an error handler is still able to handle silenced errors, but they can know whether what they handle is silenced or no.

And well-behaving error handlers which are converting errors to exceptions would be doing it only for non-silenced errors, which is what makes the deprecation warning opt-in

Comment (ddziaduch) [reply src/PhpSpec/Matcher/ScalarMatcher.php]: Thanks. I've fixed that. But I have no idea how to test this using PhpSpec. What about integration test?

Comment (ddziaduch) [reply src/PhpSpec/Matcher/ScalarMatcher.php]: @stof can you review my changes again please? :)

**Comment by gep** [comment]
@stof @ddziaduch can we merge it since requested changes are resolved?

File: composer.json [hunk 3]
```diff
@@ -23,7 +23,7 @@
 
     "require": {
         "php":                      "^7.2, <7.4",
```
Comment (gep) [inline composer.json]: Do we need to change it to smth like `"^7.2, 7.4.0",` in order to run on PHP 7.4.0?

Comment (ciaranmcnulty) [reply composer.json]: just `^7.2 <7.5`

**Comment by ddziaduch** [comment]
> @stof @ddziaduch can we merge it since requested changes are resolved?
> Or we need to get #1266 done before?

Have no idea 🙂

File: composer.json (hunk 3 above)
Comment (gep) [reply composer.json]: I think just `^7.2` is even better.

**Comment by ahilles107** [comment]
With 7.4 release tagged - can we speedup somehow merging of this PR? Looks like all tests passed.

File: composer.json (hunk 3 above)
Comment (ciaranmcnulty) [reply composer.json]: @gep no, PHP doesn't use server (otherwise this PR wouldn't have been necessary)

File: composer.json [hunk 4]
```diff
@@ -22,8 +22,8 @@
     ],
 
     "require": {
-        "php":                      "^7.2, <7.4",
-        "phpspec/prophecy":         "^1.7",
+        "php":                      "^7.2",
```
Comment (ciaranmcnulty) [inline composer.json]:
```suggestion
        "php":                      "^7.2, <7.5",
```

File: src/PhpSpec/Matcher/ScalarMatcher.php [hunk 5]
```diff
@@ -131,6 +131,10 @@ private function getCheckerName(string $name)
         if ($expected == 'boolean') {
             return 'is_bool';
         }
+        if ($expected == 'real') {
+            @trigger_error('`beReal` matcher is deprecated, please use `beFloat`', E_USER_DEPRECATED);
```
Comment (ciaranmcnulty) [inline src/PhpSpec/Matcher/ScalarMatcher.php]: I'd rather not trigger the error
```suggestion
```

**Comment by ciaranmcnulty** [comment]
Thanks for this work @ddziaduch - if you address my suggestions this can be pushed out pretty fast

**Comment by ddziaduch** [comment]
@ciaranmcnulty it's done 🙂

**Comment by ciaranmcnulty** [comment]
Oh I misunderstood and thought that this was putting the 7.4 tests into the Travis grid

**Comment by ciaranmcnulty** [comment]
Looks like 7.4 Travis on master is failing, I'll take a look later today

**Pull request merged**